enterprise/internal/authz/azuredevops/provider.go

```diff
@@ -0,0 +1,248 @@
+package azuredevops
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/sourcegraph/log"
+	authztypes "github.com/sourcegraph/sourcegraph/enterprise/internal/authz/types"
+	"github.com/sourcegraph/sourcegraph/enterprise/internal/licensing"
+	"github.com/sourcegraph/sourcegraph/internal/authz"
+	"github.com/sourcegraph/sourcegraph/internal/database"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/auth"
+	"github.com/sourcegraph/sourcegraph/internal/extsvc/azuredevops"
+	"github.com/sourcegraph/sourcegraph/internal/types"
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+var mockServerURL string
+
+func NewAuthzProviders(db database.DB, conns []*types.AzureDevOpsConnection) *authztypes.ProviderInitResult {
+	orgs, projects := map[string]struct{}{}, map[string]struct{}{}
+
+	initResults := &authztypes.ProviderInitResult{}
+
+	if len(conns) == 0 {
+		return initResults
+	}
+
+	// Iterate over all Azure Dev Ops code host connections to make sure we sync permissions for all
+	// orgs and projects in every permissions sync iteration.
+	for _, c := range conns {
+		// The list of orgs and projects may have duplicates if there are multiple Azure DevOps code
+		// host connections that have the same project in their config.
+		//
+		// Add them to a map so that we may filter out duplicates before passing them over to the
+		// provider.
+		for _, name := range c.Orgs {
+			orgs[name] = struct{}{}
+		}
+
+		for _, name := range c.Projects {
+			projects[name] = struct{}{}
+		}
+	}
+
+	// Nothing fancy, just some data restructuring to convert the map back to a slice now that we
+	// have a unique list of orgs and projects.
+	uniqueOrgs, uniqueProjects := []string{}, []string{}
+	for name := range orgs {
+		uniqueOrgs = append(uniqueOrgs, name)
+	}
+
+	for name := range projects {
+		uniqueProjects = append(uniqueProjects, name)
+	}
```

Comment:
You can now do this with [`maps.Keys`](https://pkg.go.dev/golang.org/x/exp/maps#Keys). Which is more convenient than writing a for loop :)
```suggestion
	// Convert the map back to a slice now that we have a unique list of orgs and projects.
	uniqueOrgs := maps.Keys(orgs)
	uniqueProjects := maps.Keys(projects)
```

Reply:
This is nice. TIL.

Reply:
@kopancek Can you please elaborate what you mean by does not work as expected?

Reply:
Nevermind, it does work. The problem was between my chair and my screen, as usual 🤦

Reply:
Ah. The famed Operator Error. 😛